glusterfs service: a few fixes and improvements

nh2 opened the pull request:
###### Motivation for this change

These commits fix issues I found in my production cluster; see their commit messages for individual motivations.

These are the changes I promised in https://github.com/NixOS/nixpkgs/pull/27340#discussion_r139311268.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [X] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

FYI @bachp who's another user of this service.

CC @joachifm

nh2 committed:
glusterfs service: Don't make it a prerequisite of network-online.target.

This introduces dependency cycles.

A network file system to be running is not required for a network
connection to be available.

https://github.com/NixOS/nixpkgs/commit/19759cfeab0b749666dafdf52a0aad79123a2126#commitcomment-22044519

nh2 committed:
glusterfs service: Add settings to disable rpcbind and the events daemon.

See also https://github.com/NixOS/nixpkgs/pull/22225#pullrequestreview-26459886

nh2 committed:
glusterfs service: Add killMode and stopKillTimeout options

nh2 committed:
glusterfs service: Ensure log directory exists for glustereventsd.

Prevents glustereventsd failing at startup in case it starts
before glusterd has started (whose `preStart` would also
create the needed directory).

globin commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -41,6 +41,18 @@ in
 
+      useRpcbind = mkOption {
+        type = types.bool;
+        description = "Enable use of rpcbind. This is required for Gluster's NFS functionality. You may want to turn it off to reduce the attack surface for DDoS reflection attacks. See https://davelozier.com/glusterfs-and-rpcbind-portmap-ddos-reflection-attacks/ and https://bugzilla.redhat.com/show_bug.cgi?id=1426842 for details.";
```
Could you wrap this line?

nh2 replied:
@globin Done, also for the other commit.

bachp commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -41,6 +41,53 @@ in
+          set this to 'control-group'.
+        '';
+        default = "process";
+      };
```
Does this need to be configurable? If yes give a quick description in which cases the different modes might be useful.

nh2 replied:
@bachp Updated; I have now described both typical use cases.

I have also added a commit that changes the default to `control-group`, as you and me both think that this is the better default (from https://github.com/NixOS/nixpkgs/pull/27340#discussion_r139311268).

nh2 committed:
glusterfs service: Change default killMode to "control-group".

This is a better default for NixOS because it ensures that config
changes happen fully when NixOS users expect it.

nh2 commented:
Copying into main conversation so it doesn't get hidden by `show outdated`:

I have also added a commit that changes the default to `control-group`, as you (@bachp) and me both think that this is the better default (from https://github.com/NixOS/nixpkgs/pull/27340#discussion_r139311268).

nh2 commented:
This is good to go from my side.

Once this is merged, we should switftly go on to #29062, so that we can have a good gluster 3.12 experience on NixOS 17.09.

bachp reviewed (approved):
LGTM
